addon/-private/validators.js

```diff
@@ -64,8 +64,15 @@ export function createShapeValidator(validators) {
       return error;
     }
 
+    // Ensure the value is non-null so that we won't accidentally apply get to a null value
+    if (value === null) {
+      return [`Expected a non-null object but received null`, context];
+    }
+
     for (const key of Object.keys(validators)) {
-      const error = ensureValidator(validators[key])(get(value, key), context(key));
+      // Clone the current context to generate the correct context in the next iteration
+      const nextContext = createContextPath(context());
+      const error = ensureValidator(validators[key])(get(value, key), nextContext(key));
```

Comment:
Might be good to create some kind of helper to handle this context logic... or make context smarter to handle these cases. I don't think its necessary at this moment but I believe this is the same logic we have in the array validator and I imagine we will need this any time we iterate over something where context is involved.

Reply:
Discussed offline and it might be good to create a helper `context.clone()` to clone the context. but would not add it in this PR yet.